File: ts-playwright-cucumber/src/steps/inventorySteps.ts
import { DataTable, Given, When } from '@cucumber/cucumber';
import { InventoryPage } from '../pages/inventoryPage.ts';
import type { SauceDemoWorld } from '../world.ts';

When('I add {string} to the cart', async function (this: SauceDemoWorld, productName: string) {
  await new InventoryPage(this.page).addToCart(productName);
});

When('I add the following products to the cart:', async function (this: SauceDemoWorld, table: DataTable) {
  const inventoryPage = new InventoryPage(this.page);

  for (const row of table.hashes()) {
    await inventoryPage.addToCart(row.ProductName ?? '');
  }
});

When('I go to the cart', async function (this: SauceDemoWorld) {
  await new InventoryPage(this.page).openCart();
});

Given('my cart is empty', async function (this: SauceDemoWorld) {
  await new InventoryPage(this.page).clearCart();
});

File: java-selenium-junit5-cucumber/src/test/java/com/saucedemo/uitests/steps/InventorySteps.java
package com.saucedemo.uitests.steps;

import com.saucedemo.uitests.hooks.ScenarioContext;
import com.saucedemo.uitests.pages.InventoryPage;
import io.cucumber.datatable.DataTable;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.When;
import java.util.Map;

public class InventorySteps {

    private final ScenarioContext context;

    public InventorySteps(ScenarioContext context) {
        this.context = context;
    }

    @When("I add {string} to the cart")
    public void iAddToTheCart(String productName) {
        new InventoryPage(context.driver()).addToCart(productName);
    }

    @When("I add the following products to the cart:")
    public void iAddTheFollowingProductsToTheCart(DataTable table) {
        InventoryPage inventoryPage = new InventoryPage(context.driver());

        for (Map<String, String> row : table.asMaps()) {
            inventoryPage.addToCart(row.get("ProductName"));
        }
    }

    @When("I go to the cart")
    public void iGoToTheCart() {
        new InventoryPage(context.driver()).openCart();
    }

    @Given("my cart is empty")
    public void myCartIsEmpty() {
        new InventoryPage(context.driver()).clearCart();
    }
}
